apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/profile/ProfileTaskExecutionService.java

```diff
@@ -0,0 +1,207 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.agent.core.profile;
+
+import org.apache.skywalking.apm.agent.core.boot.BootService;
+import org.apache.skywalking.apm.agent.core.boot.DefaultImplementor;
+import org.apache.skywalking.apm.agent.core.boot.DefaultNamedThreadFactory;
+import org.apache.skywalking.apm.agent.core.logging.api.ILog;
+import org.apache.skywalking.apm.agent.core.logging.api.LogManager;
+import org.apache.skywalking.apm.network.constants.ProfileConstants;
+import org.apache.skywalking.apm.util.StringUtil;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Profile task executor, use {@link #addProfileTask(ProfileTask)} to add a new profile task.
+ *
+ * @author MrPro
+ */
+@DefaultImplementor
+public class ProfileTaskExecutionService implements BootService {
+
+    private static final ILog logger = LogManager.getLogger(ProfileTaskExecutionService.class);
+
+    // add a schedule while waiting for the task to start or finish
+    private final static ScheduledExecutorService PROFILE_TASK_SCHEDULE = Executors.newSingleThreadScheduledExecutor(new DefaultNamedThreadFactory("PROFILE-TASK-SCHEDULE"));
+
+    // last command create time, use to next query task list
+    private volatile long lastCommandCreateTime = -1;
+
+    // current processing profile task context
+    private final AtomicReference<ProfileTaskExecutionContext> taskExecutionContext = new AtomicReference<>();
+
+    // profile task list, include running and waiting running tasks
+    private final List<ProfileTask> profileTaskList = Collections.synchronizedList(new LinkedList<>());
+
+    /**
+     * get profile task from OAP
+     * @param task
+     */
+    public void addProfileTask(ProfileTask task) {
+        // update last command create time
+        if (task.getCreateTime() > lastCommandCreateTime) {
+            lastCommandCreateTime = task.getCreateTime();
+        }
+
+        // check profile task limit
+        final String dataError = checkProfileTaskSuccess(task);
+        if (dataError != null) {
+            logger.warn("check command error, cannot process this profile task. reason: {}", dataError);
+            return;
+        }
+
+        // add task to list
+        profileTaskList.add(task);
+
+        // check task start now or make a schedule
+        long timeFromStartMills = task.getStartTime() - System.currentTimeMillis();
+        if (timeFromStartMills < 0) {
```

Comment:
I think we don't need this part. All tasks should be in the `PROFILE_TASK_SCHEDULE`. Could you share why we need this?

Reply:
If sniffer restart, It will also get the task from OAP, maybe some task before the current time. I used to think that the schedule could not pass in negative numbers, and I found that there would be no errors. resolved.